### PR TITLE
Add support for user defined warnings via the Warn type class

### PR DESCRIPTION
**Pull request by LiamGoodacre**
For https://github.com/purescript/purescript/issues/2564

### CODE REVIEW

File: src/Language/PureScript/Docs/Prim.hs [hunk 1]
```diff
@@ -225,6 +226,14 @@ fail = primClass "Fail" $ T.unlines
   , "[the Custom Type Errors guide](https://github.com/purescript/documentation/blob/master/guides/Custom-Type-Errors.md)."
   ]
 
+warn :: Declaration
+warn = primClass "Warn" $ T.unlines
+  [ "The Warn type class allows a custom compiler warning to be displayed."
+  , ""
+  , "For more information, see"
+  , "[the Custom Type Errors guide](https://github.com/purescript/documentation/blob/master/guides/Custom-Type-Errors.md)."
```
Comment (LiamGoodacre) [inline src/Language/PureScript/Docs/Prim.hs]: I'm currently writing a quick overview of the feature for this page - I'll submit a PR to the docs repo when this one is merged.

**Comment by LiamGoodacre** [comment]
Example output:
```
Compiling Main
[1/1 UserDefinedWarning] src/Main.purs:10:1

  10  bar :: Int
      ^^^^^^^^^^

  A custom warning occurred while solving type class constraints:

    "Bananas"

  in value declaration bar

           Src   Lib   All
Warnings   1     0     1
Errors     0     0     0
* Build successful.
```

Note that the message isn't printing in the right format due to https://github.com/purescript/purescript/issues/2568

**Comment by damncabbage** [comment]
Nice!

File: src/Language/PureScript/Docs/Prim.hs (hunk 1 above)
Comment (LiamGoodacre) [inline src/Language/PureScript/Docs/Prim.hs]: https://github.com/LiamGoodacre/documentation/blob/feature/user-defined-warnings/guides/Custom-Type-Errors.md

**Comment by paf31** [comment]
@LiamGoodacre Does this fix #2564?

**Pull request merged**